pkg/admission/permissionclaims/mutating_permission_claims.go

```diff
@@ -0,0 +1,152 @@
+
+	lcluster, err := genericapirequest.ClusterNameFrom(ctx)
+	if err != nil {
+		return err
+	}
+
+	bindings, err := m.apiBindingsIndexer.ByIndex(byWorkspaceIndex, lcluster.String())
+	if err != nil {
+		return err
+	}
+
+	for _, b := range bindings {
+		binding, ok := b.(*apisv1alpha1.APIBinding)
+		if !ok {
+			return fmt.Errorf("expected a certain type")
+		}
+		// Get the export,
+		exports, err := m.apiExportIndexer.ByIndex(byWorkspaceIndex, binding.Spec.Reference.Workspace.Path)
+		if err != nil {
+			return err
+		}
+		for _, e := range exports {
+			export, ok := e.(*apisv1alpha1.APIExport)
+			if !ok {
+				return fmt.Errorf("expected a certain typet")
+			}
+			for _, pc := range export.Spec.PermissionClaims {
+				if pc.Group == a.GetKind().Group && pc.Resource == a.GetResource().Resource {
+					key, label, err := permissionclaims.PermissionClaimToLabel(pc)
+					if err != nil {
+						return err
+					}
+					u, ok := a.GetObject().(metav1.Object)
+					if !ok {
+						return fmt.Errorf("expected type")
+					}
+					l := u.GetLabels()
+					if l == nil {
+						l = map[string]string{}
+					}
+					l[key] = label
+					u.SetLabels(l)
+				}
+			}
+		}
+	}
+
+	// Get the bindings for the given cluster.
+	// Get the exports for the bindings, and determine if any of permission claims are for the given resource.
+	// if they are, label with the internal label.
+	return nil
+}
+
+// SetKcpInformers implements the WantsExternalKcpInformerFactory interface.
+func (m *mutatingPermissionClaims) SetKcpInformers(f kcpinformers.SharedInformerFactory) {
+	if _, found := f.Apis().V1alpha1().APIBindings().Informer().GetIndexer().GetIndexers()[byWorkspaceIndex]; !found {
+		if err := f.Apis().V1alpha1().APIBindings().Informer().AddIndexers(cache.Indexers{
+			byWorkspaceIndex: func(obj interface{}) ([]string, error) {
+				return []string{logicalcluster.From(obj.(metav1.Object)).String()}, nil
+			},
+		}); err != nil {
+			// nothing we can do here. But this should also never happen. We check for existence before.
+			klog.Errorf("failed to add indexer for APIBindings: %v", err)
+		}
+	}
+	if _, found := f.Apis().V1alpha1().APIExports().Informer().GetIndexer().GetIndexers()[byWorkspaceIndex]; !found {
+		if err := f.Apis().V1alpha1().APIExports().Informer().AddIndexers(cache.Indexers{
+			byWorkspaceIndex: func(obj interface{}) ([]string, error) {
+				return []string{logicalcluster.From(obj.(metav1.Object)).String()}, nil
+			},
+		}); err != nil {
+			// nothing we can do here. But this should also never happen. We check for existence before.
+			klog.Errorf("failed to add indexer for APIBindings: %v", err)
+		}
+	}
+	m.apiBindingsIndexer = f.Apis().V1alpha1().APIBindings().Informer().GetIndexer()
+	m.apiExportIndexer = f.Apis().V1alpha1().APIExports().Informer().GetIndexer()
+	m.bindingReady = f.Apis().V1alpha1().APIBindings().Informer().HasSynced
+	m.exportReady = f.Apis().V1alpha1().APIExports().Informer().HasSynced
```

Comment:
APIExports shouldn't be necessary, and they would immediately break in a sharded setup.